packages/react-admin/src/Admin.js

```diff
@@ -29,10 +29,10 @@ const Admin = ({
     dashboard,
     history,
     locale,
-    messages = {},
     menu = Menu,
     catchAll,
     dataProvider,
+    i18nProvider = locale => this.props.messages[locale],
```

Comment:
please extract this function to a `defaultI18nProvider` to avoid redefining a new function (and therefore failing the prop identity check) whenever the `<Admin>` component is rerendered.